sourcefiles/ctoptions.py

```diff
@@ -0,0 +1,604 @@
+        
+        for idx, button in enumerate(data):
+            ret[idx] = button
+                
+        return ret
+
+    #updates button bindings given bytearray; bytearray assumed to be ordered in the same way as out
+    def update_from_bytes(self, bytes: bytearray):
+
+        if not (8 <= len(bytes) <= 11):
+            return
+
+        bytes = bytes[-8:]
+
+        for idx, byte in zip(ActionMap, bytes):
+            for input in InputMap:
+                if byte == input:
+                    self._mappings[idx] = input
+                    break
+
+
+    #restore vanilla settings, in place
+    def reset_to_vanilla(self):
+        self._mappings = self.get_vanilla()
+
+    #do not call directly; helper function to mimic functionality observed when rebinding buttons in-game
+    def _replace_overlap(self, button: InputMap, cmd: ActionMap):
+        '''
+        CT uses bitmasks stored in a known order to implement button mapping.
+        The action is controlled by the index in the array.
+        The button to which the action is bound is controlled by the value of the indexed byte
+        
+        During vblank (and inside an NMI), the functions at 0x0284EC (rearranger) and 0x028545 (translator)
+        rearrange the raw data from the SNES's joypad registers such that all buttons are on the low byte and all
+        d-pad input is on the high byte, and additional addresses are written according to the user's bindings.
+        These addresses are referenced depending on what the game needs to check for, actions or specific buttons,
+        and whether the game needs to check for a held button, a pressed button that requires releasing before
+        accepting more input, or how long the user has pressed the button.
+        
+        The Start button is never allowed to be rebound. All eight actions will still need to be mapped, thus
+        one button will always trigger two actions (or conversely, two actions will always have the same button bound to them)
+        
+        CT uses four copies of the button mapping in RAM at any one time:
+            The in-use copy, used by 0x028585 and therefore everywhere that respects user binds; 0x7E2993
+            
+            The persistent user copy, saved from RAM to SRAM when the user saves the game; 0x7E0408
+            
+            A cache of the persistent user copy, exists when user is editing binds ; 0x7E9890
+            
+            User proposed copy, edited when user is rebinding and checked against the cached copy
+            to prevent undesired lack of actions being bound to a button. ; 0x7E0f00
+                
+        To mimic reference functions at 0x02C5C7 (checks for overlap):
+        
+        Save the original button of target action
+        Iterate over array, replace actions which match target action with original button.
+        Write target button to target action
+        '''
+        
+        data = self._mappings
```

Comment:
This won't allow remapping dash == cancel to any other button configs.

Reply:
Dash is not always mirrored to Cancel; it is possible to rebind them separately. The qty (2) actions having the same button assigned to them is a consequence of CT not allowing Start to be assigned to an action.

Vod demonstrating such: https://www.youtube.com/watch?v=HR1zHiSD2PY

Reply:
You're right about dash and cancel not always being mirrored.  My mistake!

That said, there are still some issues here.  For one, I can only choose the double-mapped button by setting all of the other buttons.  An example:  Let's say I want to map dash and confirm to the Y button.  In theory there should be no problem with that from a gameplay perspective (although CT will not allow it).  If make these selections (confirm -> Y, dash -> Y), then I don't get confirm and dash on Y.  Even worse, I get confirm and cancel both mapped to B, which should be an impossible configuration.

A separate but related issue is that CT allows mapping of L/R.  Some vanilla CT runners will map menu to L or R to save anywhere/fadeout glitch past guardian.  You can't do it with default controls because of the fight triggered by the X button.